crates/pink/src/contract.rs

```diff
@@ -16,17 +16,44 @@ pub struct ExecError {
     message: String,
 }
 
+#[derive(Debug)]
+struct Hooks {
+    on_block_end: Option<u32>,
+}
+
+#[derive(Debug)]
 pub struct Contract {
     pub address: AccountId,
+    hooks: Hooks,
 }
 
 impl Contract {
     pub fn new_storage() -> Storage {
         Storage::new(Default::default())
     }
 
-    pub fn from_address(address: AccountId) -> Self {
-        Contract { address }
+    pub fn from_address(address: AccountId, storage: &mut Storage) -> Result<Self, ExecError> {
+        let (contract, _effects) = storage.execute_with(false, move || -> Result<_, ExecError> {
+            let result = Contracts::bare_call(
+                Default::default(),
+                address.clone(),
+                0,
+                GAS_LIMIT * 2,
+                Some("pink_on_block_end_selector"),
```

Comment:
Is it used to return the block end selector? If so, is it possible to just make the selector a constant value?

Reply:
No way to extract a `const` value from wasm binary given it's name. I think you meant `static`.

Actually, I originaly implement it with a global static variable. And find it is far more difficult to be done:

- The rust static vars are initialized via some sort of constructor code at runtime. So we have to load the wasm with a  real `pallet-contract` runtime to get it ready.
-  The initialized rust static vars in wasm is just a `pointer` even when the value it self is of type `u32`. So we have to read the vm's memory to extract the content of the var. This requires us to **1.** modify more of `pallet-contract` to export more make-no-sense APIs. **2.** Only endianess should be considered for `u32` but we need to consider more for more complex types.

Reply:
I meant something more naive: require a hard code name for the on_block_end function. In this way, maybe our macro can just modify the function name to something like "__pink_internal__on_block_end", and the selector will be predefined. But I'm not sure what will happen if the selector doesn't exist (like a standard ink binary).

Reply:
The selector is just a `u32` and easy to clash with a normal ink message.
If a contract do not define the `on_block_end` but have a message clashing with it. Then the runtime will call the *fake* `on_block_end` regularly。

Reply:
Reimplemented via ink events.